include/strprintf.h

```diff
@@ -6,55 +6,18 @@
 
 namespace newsboat {
 
-class StrPrintf {
-public:
-	static std::pair<std::string, std::string> split_format(
+namespace strprintf {
+	std::pair<std::string, std::string> split_format(
 		const std::string& printf_format);
 
-	static std::string fmt(const std::string& format)
-	{
-		char buffer[1024];
-		std::string result;
-		// Empty string is a dummy value that we pass in order to
-		// silence Clang's warning about format not being a literal.
-		//
-		// The thing is, at this point we know *for sure* that the
-		// format either contains no formats at all, or only escaped
-		// percent signs (which don't require any additional arguments
-		// to snprintf). It's just the way fmt recurses. The only reason
-		// we're calling snprintf at all is to process these escaped
-		// percent signs, if any. So we don't need additional
-		// parameters.
-		unsigned int len = 1 +
-			snprintf(buffer, sizeof(buffer), format.c_str(), "");
-		if (len <= sizeof(buffer)) {
-			result = buffer;
-		} else {
-			std::unique_ptr<char> buf(new char[len]);
-			snprintf(buf.get(), len, format.c_str(), "");
-			result = buf.get();
-		}
-		return result;
-	}
+	std::string fmt(const std::string& format);
 
-	template<typename... Args>
-	static std::string fmt(const std::string& format,
-		const std::string& argument,
-		Args... args)
-	{
-		return fmt(format, argument.c_str(), args...);
-	}
-
-	template<typename... Args>
-	static std::string fmt(const std::string& format,
-		const std::string* argument,
-		Args... args)
-	{
-		return fmt(format, *argument, args...);
-	}
+	template<typename T, typename... Args> std::string fmt(const std::string& format, const T& argument, Args... args);
```

Comment:
Just curious: what are the benefits of splitting a function defined in a header into a declaration and definition parts? The only reason I can see is grouping declarations into a single block so it's easier to figure out the interface by reading the header. Is that it? Are there anything else to it?

Reply:
Its so the bodies of the proceeding template functions can be aware of the existance of the other template functions.
I guess we dont need the forward declaration of whatever the first template function is, but I think more consistent to just have all of them.